[SYCL] Correctly handle debug information in global offset pass

jchlanda opened the pull request:
Make sure that cloned functions do not use debug information that points to the original function. In doing so for each cloned function, create a `DISubprogram` that describe the clone and walk the instructions of cloned functions to correctly fix up the debug information.

jchlanda commented:
>     2. Looking at the change itself, this issue could arise whenever anyone tries to clone a function. Can we address this upstream whenever we clone a function?

You are right, I think if we manage to populate the `ValueToValueMapTy &VMap` of the `CloneFunctionInto` with all the duplicated functions, then llvm would resolve the debug information for us, and the manual fixups done in this pass will not be needed.
I've marked this PR as a draft and will explore the `VMap` solution.

Thank you.

asudarsa commented:
Hi @jchlanda 

Can you please let me know about the recent changes on this PR? I see only one commit and am unable to comment about the changes after my earlier comment.

Thanks

jchlanda commented:
> Hi @jchlanda
> 
> Can you please let me know about the recent changes on this PR? I see only one commit and am unable to comment about the changes after my earlier comment.
> 
> Thanks

HI @asudarsa 

I've followed your suggestion and worked on a solution that utilizes llvm's `CloneFunction` machinery. Essential what this patch does now is that it creates all the functions that would be cloned during the execution of the pass ahead of time - that is before we get to create functions with offset. This is done so that we can populate a value-to-value map that llvm uses to keep track of all the debug information amongst other).

While this PR was in draft state, I rebased it and dropped some of the earlier, now not needed, commits.

Thanks.

jchlanda commented:
Ping @intel/dpcpp-tools-reviewers this PR should now be ready for your review.
Thank you.

jchlanda commented:
Friendly ping @intel/dpcpp-tools-reviewers

npmiller commented:
@intel/dpcpp-tools-reviewers could we get some help with an approval/review on this? It's actually only needed for CUDA and HIP (maybe we should change the codeowners on these files?), and we'd like to get this patch into the next release.

npmiller commented:
Updated the codeowners here: https://github.com/intel/llvm/pull/17484

jchlanda commented:
With the change to the [codeowners structure](https://github.com/intel/llvm/pull/17484) and @npmiller's approval this should be read to merge @intel/llvm-gatekeepers

sommerlukas commented:
@jchlanda Github UI still says that merge is blocked due to missing review from @intel/dpcpp-tools-reviewers, so the change to code owners might only take effect for new PRs.

@intel/dpcpp-tools-reviewers can we get a quick review on this one?

jchlanda commented on llvm/lib/SYCLLowerIR/GlobalOffset.cpp:
```diff
@@ -91,6 +92,70 @@ static void validateKernels(Module &M, TargetHelpers::KernelCache &KCache) {
 }
 
+void GlobalOffsetPass::createClonesAndPopulateVMap(
+    TargetHelpers::KernelCache &KCache, Function *ImplicitOffsetIntrinsic) {
```
Yeap, done.

jchlanda commented on llvm/lib/SYCLLowerIR/GlobalOffset.cpp:
```diff
@@ -91,6 +92,70 @@ static void validateKernels(Module &M, TargetHelpers::KernelCache &KCache) {
+    Arguments.push_back(ImplicitArgumentType);
+
+    // Build the new function.
+    assert(!FuncTy->isVarArg() && "Variadic arguments prohibited in SYCL");
```
This is an old code that I've shuffled around in this PR, but I take your point, changing it to `llvm_unreachable` instead.

maarquitos14 reviewed (approved):
LGTM.

jchlanda commented:
@intel/llvm-gatekeepers we should be ready to roll now. Thank you.
